pkg/docadriverimages/doca_drivers.go

```diff
@@ -0,0 +1,141 @@
+/*
+2024 NVIDIA CORPORATION & AFFILIATES
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package docadriverimages package provides information about DOCA driver images
+package docadriverimages
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	mellanoxv1alpha1 "github.com/Mellanox/network-operator/api/v1alpha1"
+	"github.com/Mellanox/network-operator/pkg/config"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kauth "github.com/google/go-containerregistry/pkg/authn/kubernetes"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Provider provides interface to check the DOCA driver images
+type Provider interface {
+	// TagExists returns true if DOCA driver image with provided tag exists
+	TagExists(tag string) bool
+	// SetImageSpec sets the Container registry details
+	SetImageSpec(*mellanoxv1alpha1.ImageSpec)
+}
+
+// NewProvider creates a provider for  DOCA driver images,
+// queries the container image registry to get the exiting tags
+func NewProvider(ctx context.Context, c client.Client) Provider {
+	p := &provider{c: c, docaImageSpec: nil, tags: make([]string, 0), ctx: ctx}
+
+	ticker := time.NewTicker(time.Duration(config.FromEnv().State.DocaDriverImagePollTimeMinutes) * time.Minute)
+
+	go func() {
+		for ; ; <-ticker.C {
+			p.retrieveTags()
+		}
+	}()
+	return p
+}
+
+// provider is a static implementation of the Provider interface
+type provider struct {
+	c             client.Client
+	tags          []string
+	docaImageSpec *mellanoxv1alpha1.ImageSpec
+	ctx           context.Context
+	mu            sync.Mutex
+}
+
+// TagExists returns true if DOCA driver image with provided tag exists
+func (p *provider) TagExists(tag string) bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	for _, t := range p.tags {
+		if t == tag {
+			return true
+		}
+	}
+	return false
+}
+
+// SetImageSpec sets the Container registry details
+func (p *provider) SetImageSpec(spec *mellanoxv1alpha1.ImageSpec) {
+	if spec == nil {
```

Comment:
I think we need to protect this method with mutex as well. 

retrieveTags is running in a background mode and  in theory line 111 can panic if SetImageSpec was called between check on line 103 and loop on line 111

Reply:
I added the mutex.
Since it calls, `retrieveTags` the mutex needs to be explicitly release, and not in a `defer` call